orderer/common/follower/block_puller.go

```diff
@@ -24,73 +24,152 @@ import (
 	}
 	stdDialer.Config.AsyncConnect = false
 	stdDialer.Config.SecOpts.VerifyCertificate = nil
 
-	// Extract the TLS CA certs and endpoints from the join-block
-	endpoints, err := cluster.EndpointconfigFromConfigBlock(joinBlock, bccsp)
-	if err != nil {
-		return nil, errors.Wrap(err, "error extracting endpoints from config block")
-	}
-
 	der, _ := pem.Decode(stdDialer.Config.SecOpts.Certificate)
 	if der == nil {
 		return nil, errors.Errorf("client certificate isn't in PEM format: %v",
 			string(stdDialer.Config.SecOpts.Certificate))
 	}
 
+	factory := &BlockPullerCreator{
+		channelID: channelID,
+		bccsp:     bccsp,
+		blockSigVerifierFactory: &cluster.BlockVerifierAssembler{
+			Logger: logger,
+			BCCSP:  bccsp,
+		},
+		clusterConfig:       clusterConfig,
+		signer:              signer,
+		stdDialer:           stdDialer,
+		der:                 der,
+		clusterVerifyBlocks: cluster.VerifyBlocks, // The default block sequence verification method.
+	}
+
+	if len(alternativeVerifyBlocksFunc) > 0 { // Provide an alternative, mainly for tests.
+		factory.clusterVerifyBlocks = alternativeVerifyBlocksFunc[0]
+	}
+
+	return factory, nil
+}
+
+// BlockPuller creates a block puller on demand.
+func (factory *BlockPullerCreator) BlockPuller(configBlock *common.Block) (ChannelPuller, error) {
+	// Extract the TLS CA certs and endpoints from the join-block
+	endpoints, err := cluster.EndpointconfigFromConfigBlock(configBlock, factory.bccsp)
+	if err != nil {
+		return nil, errors.Wrap(err, "error extracting endpoints from config block")
+	}
+
 	bp := &cluster.BlockPuller{
-		VerifyBlockSequence: verifyBlockSequence,
-		Logger:              flogging.MustGetLogger("orderer.common.cluster.puller").With("channel", channelID),
-		RetryTimeout:        clusterConfig.ReplicationRetryTimeout,
-		MaxTotalBufferBytes: clusterConfig.ReplicationBufferSize,
-		FetchTimeout:        clusterConfig.ReplicationPullTimeout,
+		VerifyBlockSequence: factory.VerifyBlockSequence,
+		Logger:              flogging.MustGetLogger("orderer.common.cluster.puller").With("channel", factory.channelID),
+		RetryTimeout:        factory.clusterConfig.ReplicationRetryTimeout,
+		MaxTotalBufferBytes: factory.clusterConfig.ReplicationBufferSize,
+		FetchTimeout:        factory.clusterConfig.ReplicationPullTimeout,
 		Endpoints:           endpoints,
-		Signer:              signer,
-		TLSCert:             der.Bytes,
-		Channel:             channelID,
-		Dialer:              stdDialer,
+		Signer:              factory.signer,
+		TLSCert:             factory.der.Bytes,
+		Channel:             factory.channelID,
+		Dialer:              factory.stdDialer,
 	}
 
 	return bp, nil
 }
+
+// UpdateVerifierFromConfigBlock creates a new block signature verifier from the config block and updates the internal
+// link to said verifier.
+func (factory *BlockPullerCreator) UpdateVerifierFromConfigBlock(configBlock *common.Block) error {
+	configEnv, errC := cluster.ConfigFromBlock(configBlock)
```

Comment:
Style nit: why have errC and errV here instead of just err? Doesn't seem to serve a purpose.

Reply:
done